045-kafka-roller-refactoring.md

```diff
@@ -0,0 +1,100 @@
+- Large clusters: duration of rolling restarts becomes a problem.
+- Tension with Cruise Control e.g. over leadership.
+
+## Motivation
+
+Refactoring the Kafka Roller is a change that needs to be well planned and executed since the KafkaRoller is a crucial and complex component of the Cluster Operator.
+
+That is why this proposal aims for:
+- Proposing the desired state of the Kafka Roller.
+- Proposing a safe implementation path that focuses on incremental small changes which lead to achieving such desired state.
+
+## Proposal
+
+### Enhancement Areas
+
+The proposal focuses on the following areas of enhancements
+
+[//]: # (TODO: Change name)  
+
+#### Observations
+
+* Kafka Roller decisions can become more accurate and better informed by having rich observations coming from different sources (e.g. Kafka admin API, Kubernetes API, metrics, etc.).
+* Creating an abstraction for the observations' collection that encapsulates different observation sources, including custom sources (e.g. a Kafka Agent). Such custom sources would help with better understanding of the Kafka server state (e.g. broker is in log recovery).
+
+#### State Classification
+
+* Kafka servers can be classified into a set of states that are determined by the collected observations (e.g. needs restart, unhealthy, stable, etc.)
+* The Kafka server states can be modeled as a state machine which reflects the Kafka Roller's understanding of the Kafka server state and determines the possible actions needed to transition the Kafka server to a healthy state.
+* The Kafka Roller can continuously collect observations, classify the Kafka server state, and reconciling Kafka servers in unhealthy states via the defined state machine transitions.
+
+#### Decoupling Observations' Collection, Classification and Pod Management
+
+A new set of abstractions can be introduced around observations' collection, classification and pod management that would decouple those three aspects for the following reasons:   
+* Breaking down the complexity of the Kafka Roller and making it easier to reason about.
+* Better testability of each aspect of the roller in isolation.
+* Easier extensibility of the Kafka Roller to support future scenarios (e.g. different sources of observations, different classification logic, etc.)
+
+### Proposed Abstractions
+
+![](./images/045-kafka-roller-refactoring-design.png)
+
+#### The Observer
+
+* An abstraction around observations' collections.
+* It encapsulates the logic of collecting observations per Kafka server from different sources (e.g. Kafka admin API, Kubernetes API, metrics, etc.).
+
+#### The Classification
+
+* Responsible for mapping observations to states.
+* Does not understand state transitions.
+
+#### The Processor
+
+* An abstraction around Kafka server management. 
+* Responsible for pod deletion and reconfiguration
+* Does not understand state transitions.
+
```

Comment:
You need to go into detail about how the processor works:

1. That it only processes a single pod/server at a time. Aside: With the right checks I don't think this is something we need to enforce directly; i.e. the acks=all availability constraint on its own might be enough to prevent bad things happening, without forcing the algorithm to be sequential. But starting with sequential seems a sensible starting point.
2. How does it determine which pod to process next? My protocol code used a total order on the states to determine order, if you recall?
3. That once we've attempted to restart/reconfigure a pod the first time we don't progress to another pod until we've see that pod enter the `stable` state. 
4. Under what circumstances does the algorithm terminate. Running out of servers that need reconfig/restart is the happy path, but what about exceptional paths? Having an algorithm that definitely terminates, even if via a timeout, is a requirement to avoid the operator getting stuck.